File: test/interface/golden/product/TwoTypeParameters.ml
type ('a0, 'a1) twoTypeParameters =
  { ttpId : int
  ; ttpFirst : 'a0
  ; ttpSecond : 'a1
  ; ttpThird : ('a0 * 'a1)
  }

let encodeTwoTypeParameters encodeA0 encodeA1 x =
  Aeson.Encode.object_
    [ ( "ttpId", Aeson.Encode.int x.ttpId )
    ; ( "ttpFirst", encodeA0 x.ttpFirst )
    ; ( "ttpSecond", encodeA1 x.ttpSecond )
    ; ( "ttpThird", (Aeson.Encode.pair encodeA0 encodeA1) x.ttpThird )
    ]

let decodeTwoTypeParameters decodeA0 decodeA1 json =
  match Aeson.Decode.
    { ttpId = field "ttpId" int json
    ; ttpFirst = field "ttpFirst" (fun a -> unwrapResult (decodeA0 a)) json
    ; ttpSecond = field "ttpSecond" (fun a -> unwrapResult (decodeA1 a)) json
    ; ttpThird = field "ttpThird" (pair (fun a -> unwrapResult (decodeA0 a)) (fun a -> unwrapResult (decodeA1 a))) json
    }
  with
  | v -> Js_result.Ok v
  | exception Aeson.Decode.DecodeError message -> Js_result.Error ("decodeTwoTypeParameters: " ^ message)

File: test/interface/golden/lib/js/product/TwoTypeParameters.js
'use strict';

var Block        = require("bs-platform/lib/js/block.js");
var Curry        = require("bs-platform/lib/js/curry.js");
var Js_exn       = require("bs-platform/lib/js/js_exn.js");
var Aeson_decode = require("bs-aeson/lib/js/src/Aeson_decode.js");
var Aeson_encode = require("bs-aeson/lib/js/src/Aeson_encode.js");

function encodeTwoTypeParameters(encodeA0, encodeA1, x) {
  return Aeson_encode.object_(/* :: */[
              /* tuple */[
                "ttpId",
                x[/* ttpId */0]
              ],
              /* :: */[
                /* tuple */[
                  "ttpFirst",
                  Curry._1(encodeA0, x[/* ttpFirst */1])
                ],
                /* :: */[
                  /* tuple */[
                    "ttpSecond",
                    Curry._1(encodeA1, x[/* ttpSecond */2])
                  ],
                  /* [] */0
                ]
              ]
            ]);
}

function decodeTwoTypeParameters(decodeA0, decodeA1, json) {
  var exit = 0;
  var v;
  try {
    v = /* record */[
      /* ttpId */Aeson_decode.field("ttpId", Aeson_decode.$$int, json),
      /* ttpFirst */Aeson_decode.field("ttpFirst", (function (a) {
              return Aeson_decode.unwrapResult(Curry._1(decodeA0, a));
            }), json),
      /* ttpSecond */Aeson_decode.field("ttpSecond", (function (a) {
              return Aeson_decode.unwrapResult(Curry._1(decodeA1, a));
            }), json)
    ];
    exit = 1;
  }
  catch (raw_exn){
    var exn = Js_exn.internalToOCamlException(raw_exn);
    if (exn[0] === Aeson_decode.DecodeError) {
      return /* Error */Block.__(1, ["decodeTwoTypeParameters: " + exn[1]]);
    } else {
      throw exn;
    }
  }
  if (exit === 1) {
    return /* Ok */Block.__(0, [v]);
  }
  
}

exports.encodeTwoTypeParameters = encodeTwoTypeParameters;
exports.decodeTwoTypeParameters = decodeTwoTypeParameters;
/* Aeson_encode Not a pure module */
